### PR TITLE
MSD Detector Added

### PR DESCRIPTION
**Pull request by m4nh**

### CODE REVIEW

**Comment by m4nh** [comment]
@mshabunin so now? I push a new PR on Opencv_extras with a branch "add_msd_detector" like in this PR.

**Comment by cbalint13** [comment]
@mshabunin , are there things to done for opencv_extra ? i.e. push there the related test comparasion data ? It seems this PR is failing now there.

 Let us know, i try guide @m4nh .

**Comment by mshabunin** [comment]
@m4nh, now the build system uses your _opencv_extra_ version (see SourceStamps block [here](http://master.ocv/buildbot/builders/precommit-contrib_linux64/builds/1806)), but tests are still failing. 

You have to generate regression data for [performance test](http://master.ocv/buildbot/builders/precommit-contrib_linux64/builds/1806/steps/perf_xfeatures2d/logs/stdio). Please see this article: https://github.com/Itseez/opencv/wiki/HowToUsePerfTests#how-to-update-perf-data (just run yor test with `--perf_write_sanity` flag)

For accuracy test you will need to fix you detector, because it have returned invalid keypoint: `KeyPoint::pt is out of image (x=-nan, y=-nan).`

**Comment by cbalint13** [comment]
@m4nh , will help you with this against opencv_extra repo ! Be back a bit later.

**Comment by cbalint13** [comment]
@m4nh , Looked at the issue. Please let me know if we still have buildbot failures. I would be glad to help more if the case.

* Dived a bit in the two problems remarked by the build-bot.
* I understand buildbot now is fully functional not like this summer.
* Here are proposed solutions to fix issues: http://openrisc.rdsor.ro/MSD-fixes/

1) I believe this fix bellow should be obvious, otherwise saliency array is not released:
```
@@ -236,9 +238,9 @@
 
                 nonMaximaSuppression(saliency, keypoints);
 
-                for (int r = 0; r < m_n_scales; r++)
+                for (int r = 0; r < m_cur_n_scales; r++)
                 {
-                    delete[] saliency[r];
+                    saliency[r].clear();
                 }
 
                 m_scaleSpace.clear();
```
2) A lot of "nan" pops inside the the code (i used feenableexcept() to catch them). Interesting thing is that they pops at random points so it seems to be a cast/pointer thing rather than some numerical thing. It is solved by using std::vector< std::vector float > instead of std::vector float, see http://openrisc.rdsor.ro/MSD-fixes/msd.diff . I find this quick way to solve, honestly i don't know why original float pointer array wasn't working i believe it doesn't cast properly well somewhere.

3) I notice your PR agaist "opencv_extra": https://github.com/m4nh/opencv_extra/tree/add_msd_detector , that is good and necessary. Just please remove that "testdata/stitching/flintstones.jpg" and add http://openrisc.rdsor.ro/MSD-fixes/xfeatures2d.xml to the opencv_extra repo here: "opencv_extra/testdata/perf/xfeatures2d.xml" This xfeatures2d.xml was generated just as @mshabunin told.

@StevenPuttemans , can review a bit i hope at 3 proposal is fine (not a github expert).

**Comment by StevenPuttemans** [comment]
@cbalint13 I am far from an expert myself :) but your suggestions seem correct!

**Comment by m4nh** [comment]
@cbalint13 @StevenPuttemans seems to be ok now. but i push Contrib before Extra so it fails because of order

**Comment by cbalint13** [comment]
@m4nh , make a small change somewhere in the code i.e. add/remove a newline and push again.
* Also i reference here that PR: https://github.com/Itseez/opencv_extra/pull/277

**Comment by m4nh** [comment]
@cbalint13 it fails again!

**Comment by cbalint13** [comment]
@mshabunin , can look why we still have perf failure ? You can look at https://github.com/Itseez/opencv_extra/pull/277 , it was merged/fetched properly by buildbot but we get "No regression data for points-pt argument" . That xfeatures2d.xml pass well on local machine.

Can help ?

**Comment by cbalint13** [comment]
@mshabunin , i think we got the cause.
@m4nh , can please not touch testdata/perf/features2d.xml (revert any change against that file to original one) but create a new file called "testdata/perf/xfeatures2d.xml"  (i mean for https://github.com/Itseez/opencv_extra/pull/277) ?

* At http://openrisc.rdsor.ro/MSD-fixes/ there was the new file "xfeatures2d.xml" you can use it.
* I am quite sure this is the cause for our last "No regression data for points-pt argument" error.

**Comment by m4nh** [comment]
@cbalint13  no way

**Comment by mshabunin** [comment]
@m4nh, tests fail due to accuracy differences. Please check if you really need such small eps here: `SANITY_CHECK_KEYPOINTS(points, 1e-5);`

Results are close, but not enough:
```
The difference between expect_max and actual_max is 0.00048828125, which exceeds eps, where
expect_max evaluates to 5854.92822265625,
actual_max evaluates to 5854.9287109375, and
eps evaluates to 1.0000000000000001e-05.
...
The difference between expect_min and actual_min is 1.52587890625e-05, which exceeds eps, where
expect_min evaluates to 250.39794921875,
actual_min evaluates to 250.39796447753906, and
eps evaluates to 1.0000000000000001e-05.
```

**Comment by cbalint13** [comment]
@m4nh , Daniele,

* We could try lower that 1e-5  to e.g. 1e-3. 
@mshabunin is this difference coming from -fast-math compiler thing ?

**Comment by m4nh** [comment]
thanks @cbalint13 ! So i have to change EPS in perf msd test? or i have to wait for @mshabunin answer?

**Comment by cbalint13** [comment]
@m4nh , i believe there is no harm if we change right now to 1e-3. I suggest to do it. I asked @mshabunin the root of cause for my curiosity.

Daniele, anyway MSD is really great i am really looking forward to see it merged !

**Comment by StevenPuttemans** [comment]
@m4nh Daniele you can go ahead and try to change the value. As @mshabunin stated, it might be to strict or sharp and the complete PR got be okay to go with this little change.

So lets just try it out? :D

**Comment by mshabunin** [comment]
@cbalint13 , AFAIK the fast-math option is not set during compilation on x86. The results vary because of platform/compiler floating point math differences.

**Comment by m4nh** [comment]
@cbalint13  thanks! but for MSD we have to thank @fedassa :) 
i will try now new epsilon

**Comment by fedassa** [comment]
@cbalint13 thanks, we are really eager to see MSD merged into OpenCV. Hopefully it will soon be part of the official keypoint detector set ;)

**Comment by cbalint13** [comment]
On Mon, Nov 2, 2015 at 2:31 PM, Maksim Shabunin <notifications@github.com>
wrote:

> @cbalint13 <https://github.com/cbalint13> , AFAIK the fast-math option is
> not set during compilation on x86. The results vary because of
> platform/compiler floating point math differences.
>

Maksim,

Ahh, got it ! This is the difference, i had -fast-math enabled. Good to
know the source of tiny difference :) Well, i think going lower with test
limit 10e-3 is fine too.

**Comment by m4nh** [comment]
@cbalint13  it fails with:

test_descriptors_regression.cpp:444:1 fatal error: error writing to /tmp/cc3fNTMT.s: No space left on device

**Comment by Dikay900** [comment]
just amend your last commit and force push again this will trigger the build bot again

File: modules/xfeatures2d/src/msd.cpp [hunk 1]
```diff
@@ -0,0 +1,781 @@
...
+
+[1] F. Tombari, L. Di Stefano
+"Interest Points via Maximal Self-Dissimilarities"
+12th Asian Conference on Computer Vision (ACCV), 2014
+
+The code is ported from the stand-alone implementation available
+at this repository:
+www.github.com/fedassa/msdDetector
+
+AUTHORS:  Federico Tombari (fedassa@gmail.com) (original code),
```
Comment (mshabunin) [inline modules/xfeatures2d/src/msd.cpp]: Please remove trailing space at this line.

**Comment by m4nh** [comment]
hi @cbalint13 can you confirm that the PR has been fully accepted? Still cannot see it in official repository

**Comment by cbalint13** [comment]
@m4nh ,

- I am not maintainer, but in my opinion all is fine with this PR. The quality of code and algorithm itself is exceptional, it also pass opencv's exigent requirements too. If you ask me this detector should be  really welcomed.

- For some reason bigger PR's were not merged all this summer and autumn (i guess reasons was GSOC2015 and the buildbot maintainance/upgrade), but I can see recently (~last 2 weeks) that PR's are starting to be merged in order.

- I also have two opened bigger PRs, but i do not complain, i understand the situation.

I try ping @vpisarev , in hope he will lead / help us or even merge this PR.

**Comment by StevenPuttemans** [comment]
Hello @m4nh, it seems that everything is fine now and you should indeed patiently wait in line to get this merged in. Let us keep fingers crossed that it goes smoothly.

As a last commit I would like you to ask to do a `git rebase -i HEAD~8`. Except for the first element, change the line headings to `s`, to squash commits. Then change to a single commit message, then do a `got push -f --all origin` to make sure your PR has only a single commit and does not cramp up the commit history :)

**Comment by m4nh** [comment]
@StevenPuttemans @mshabunin  i think that merge and squash together went wrong

**Comment by mshabunin** [comment]
@m4nh , I've rebased your changes, you just need to squash them: https://github.com/Itseez/opencv_contrib/compare/master...mshabunin:add_msd_detector

**Comment by m4nh** [comment]
@mshabunin sorry but i don't understand how :( i m new in using git across multiple forks on github. In my fork the squash is done. can you squash commits in your forks?

**Comment by mshabunin** [comment]
@m4nh , no, you only made a rebase, not squash. It should be easy:
```.sh
# get branch from my fork
git remote add temp https://github.com/mshabunin/opencv_contrib.git
git fetch temp add_msd_detector
# set your local branch to the same commit
git checkout temp/add_msd_detector -B add_msd_detector
# get upstream changes (upstream = Itseez/opencv_contrib)
git fetch upstream master
# do the interactive rebase
git rebase -i upstream/master
# in the editor mark all commits 'f' or 'fixup' except the first one, then save and exit
# force push to your repository (origin = m4nh/opencv_contrib)
git push -f origin add_msd_detector
# remove unnecessary remote
git remote remove temp
```
I hope I didn't miss anything

**Comment by vpisarev** [comment]
@mshabunin, let me assign this one to you; I myself is fine with the patch

**Comment by mshabunin** [comment]
Moved to #473

**Pull request closed without merging**